contracts/staking/ServiceStakingBase.sol

```diff
@@ -0,0 +1,309 @@
+        // Get the service Id set length
+        uint256 size = setServiceIds.length;
+        uint256 lastAvailableRewards = availableRewards;
+        uint256 tsCheckpointLast = tsCheckpoint;
+
+        // Number of services eligible for the reward during the current checkpoint
+        uint256 numServices;
+
+        // If available rewards are not zero, proceed with staking calculation
+        // Otherwise, just bump the timestamp of last checkpoint
+        if (serviceId > 0 && lastAvailableRewards > 0) {
+            uint256[] memory eligibleServiceIds = new uint256[](size);
+            uint256[] memory eligibleServiceRewards = new uint256[](size);
+            uint256[] memory serviceIds = new uint256[](size);
+            uint256[] memory serviceNonces = new uint256[](size);
+            uint256 totalRewards;
+
+            // Calculate each staked service reward eligibility
+            for (uint256 i = 0; i < size; ++i) {
+                // Get the current service Id
+                serviceIds[i] = setServiceIds[i];
+
+                // Get the service info
+                ServiceInfo storage curInfo = mapServiceInfo[serviceIds[i]];
+
+                // Calculate the staking nonce ratio
+                uint256 curNonce = IMultisig(curInfo.multisig).nonce();
+                // Get the last service checkpoint: staking start time or the global checkpoint timestamp
+                uint256 serviceCheckpoint = tsCheckpointLast;
+                // Adjust the service checkpoint time if the service was staking less than the current staking period
+                if (curInfo.tsStart > tsCheckpointLast) {
+                    serviceCheckpoint = curInfo.tsStart;
+                }
+                // Calculate the liveness ratio in 1e18 value
+                uint256 ratio;
+                // If the checkpoint was called in the exactly same block, the ratio is zero
+                if (block.timestamp > tsCheckpointLast) {
+                    ratio = ((curNonce - curInfo.nonce) * 1e18) / (block.timestamp - serviceCheckpoint);
+                }
+
+                // Record the reward for the service if it has provided enough transactions
+                if (ratio >= livenessRatio) {
+                    // Calculate the reward up until now and record its value for the corresponding service
+                    uint256 reward = rewardsPerSecond * (block.timestamp - serviceCheckpoint);
+                    totalRewards += reward;
+                    eligibleServiceRewards[numServices] = reward;
+                    eligibleServiceIds[numServices] = serviceIds[i];
+                    ++numServices;
+                }
+
+                // Record current service multisig nonce
+                serviceNonces[i] = curNonce;
+
+                // Record the unstaked service Id index in the global set of staked service Ids
+                if (serviceIds[i] == serviceId) {
+                    idx = i;
+                }
```

Comment:
Should not happen in this scope imo. You have a tendency to combine a lot of individual responsibilities

Reply:
Yeah, that's why I'm usually gas efficient ;)